API-compatibility with python sets.

urdvr opened the pull request:
Worked on ideas from #37
 
This pull request adds the methods:

* `isdisjoint`
* `issubset`
* `issuperset`
* `difference`
* `symmetric_difference`
* `copy`
* `difference_update`
* `symmetric_difference_update`
* `clear`
* `pop`

It changes the behavior of `union` and `intersection`. They are no longer class methods
```python
a = BitMap([1, 2])
b = BitMap([3, 4])
c = BitMap([4, 5])
a.union(b, c) #BitMap([1, 2, 3, 4, 5])
BitMap.union(a, b, c) #BitMap([1, 2, 3, 4, 5])

#NB! Despite being called from FrozenBitMap, the type of the result is taken from the first argument
FrozenBitMap.union(a,b,c) #BitMap([1, 2, 3, 4, 5])
BitMap.union(FrozenBitMap(a), b, c) #FrozenBitMap([1, 2, 3, 4, 5])

#For native sets the behaviour is even more restrictive.
set1 = set([1, 2])
set2 = set([3, 4])
set.union(set1, frozenset(set2))  #{1, 2, 3, 4}
set.union(frozenset(set1), set2)  #TypeError
``` 
The implementation of `union` in this pull request seemed to be a bit faster than the previous one in quick tests. This seems surprising and needs a closer look. It might have been an artifact of the random data I used for testing.

- [x] add doctests for new methods
- [x] benchmark previous `union` and `intersection` against new ones with more diverse data

urdvr commented:
I don't understand all the implications of `copy_on_write`, so there might be issues around that. 

I'm submitting to get feedback in case I'm totally off target in some way.

Ezibenroc commented on pyroaring/abstract_bitmap.pxi:
```diff
@@ -209,44 +209,85 @@ cdef class AbstractBitMap:
-            return (<AbstractBitMap>cls()).from_ptr(result) # FIXME to change when from_ptr is a classmethod
+        union = BitMap(self)
+        for other in others:
+            union.__ior__(other)
```
I would be surprised that this new implementation is really faster than the original one.

Here is a first quick benchmark:
```bash
python -m timeit -s 'from pyroaring import BitMap; import random; size=1000000; bitmaps=[BitMap(random.sample(range(size), size//20)) for _ in range(100)]' 'BitMap.union(*bitmaps)'
```
* new implementation: 100 loops, best of 3: 3.89 msec per loop
* old implementation: 100 loops, best of 3: 2.87 msec per loop

urdvr replied:
Yeah, it looks like it depends on the dataset. I tested on much more dense sets.

```python
import random
size = 1000000
bitmaps = []
for i in range(100):
    b = BitMap(range(size))
    for i in range(size // 20):
        b.discard(random.randint(0, size))
    b.run_optimize()
    bitmaps.append(b)
```
* new implementation: 239 µs ± 19.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
* old implementation: 480 µs ± 17.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

And the difference is much more drastic without `run_optimize`.

* new implementation: 1.33 ms ± 52.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
* old implementation: 25.3 ms ± 1.79 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

Ezibenroc commented on pyroaring/bitmap.pxi:
```diff
@@ -116,6 +116,37 @@ cdef class BitMap(AbstractBitMap):
+        """
+        union = self & other
+        self.__ior__(other)
+        self.__isub__(union)
```
You should simply call this:
```python
self.__ixor__(other)
```

Ezibenroc commented on pyroaring/bitmap.pxi:
```diff
@@ -116,6 +116,37 @@ cdef class BitMap(AbstractBitMap):
+
+    def clear(self):
+        """Remove all elements from this set."""
+        self.__isub__(self)
```
You should use [roaring_bitmap_clear](https://github.com/RoaringBitmap/CRoaring/blob/995cabc6ff179cefd930cf68c35d5ddd8df112eb/include/roaring/roaring.h#L330) function.

Ezibenroc commented:
Thanks a lot for your work!

I did not check the tests yet, but there are several things that should be changed in the code (see my comments).

urdvr commented:
Thanks for the review, I'll make the changes over the week as I find time.

lemire commented:
Regarding the bitmap aggregation, there is a note in the README about this:

https://github.com/RoaringBitmap/CRoaring#how-to-best-aggregate-bitmaps

@urdvr Your chosen benchmark is such that, quickly, the aggregated bitmap is fully dense and the rest of the aggregations could (statistically) be omitted.

I think that changing the aggregation code the way you did is fine, but you must understand that if it got faster in this case, it will get slower for other cases. We are using heuristics trained on data, trying to guess what might be typical data.

What I suggest is that you actually benchmark this on data you genuinely care about. You should not tune code based on a one-off benchmark. Use realistic data if you can.

Hope this helps.

Ezibenroc commented:
Regarding the failing tests in Python2, the issue seems to be this:

```python
class A:
    def test(self, *args):
        print(self, args)

A.test(1, 2, 3)  # succeeds in Python3, fails in Python2...
```
From what I understood, when calling an instance method as a class method, the first argument **must** be an instance of that class in Python2 (but not necessarily in Python3). Thus, this call fails: 
```python
BitMap.intersection(FrozenBitMap(), FrozenBitMap())
```

I do not know if there is an easy solution for that. To be honest, I will probably drop Python2 support soon, so feel free to disable this particular test when it is run with Python2 (but keep it for Python3).

lemire commented:
If you disable some tests on Python 2, I think it is probably best to openly state that you no longer support Python 2 going forward.

urdvr commented:
I reverted the implementation of `union`, `intersection` and changed `difference` to match `intersection.`
From benchmarking it was obvious that the original code was faster in almost all cases.

I added special cases for when the methods get called with two BitMaps as that seemed to be an universal improvement (and that's the use-case I care most about).

I'll take of the WIP tag as I consider the pull request to be feature complete now and I have no improvements planned, but I'll still respond to suggestions and comments.

Ezibenroc commented on pyroaring/abstract_bitmap.pxi:
```diff
@@ -209,8 +209,95 @@ cdef class AbstractBitMap:
+        size = len(bitmaps)
+        cdef AbstractBitMap result, bm
+        if size <= 1:
+            return bitmaps[0].__class__(*bitmaps)
```
This would be equivalent and clearer:
```python
return bitmaps[0].copy()
```

Ezibenroc replied:
This would be equivalent and clearer:
```python
return bitmaps[0].copy()
```

Ezibenroc commented on pyroaring/abstract_bitmap.pxi:
```diff
@@ -209,8 +209,95 @@ cdef class AbstractBitMap:
+        if size <= 1:
+            return bitmaps[0].__class__(*bitmaps)
+        elif size == 2:
+            return (<AbstractBitMap>bitmaps[0]).binary_op(<AbstractBitMap?>bitmaps[1], croaring.roaring_bitmap_andnot)
```
Similarly, this would be better:
```python
return bitmaps[0] - bitmaps[1]
```

Ezibenroc commented on pyroaring/abstract_bitmap.pxi:
```diff
@@ -222,16 +309,17 @@ cdef class AbstractBitMap:
         cdef vector[const croaring.roaring_bitmap_t*] buff
         if size <= 1:
-            return cls(*bitmaps)
+            return bitmaps[0].__class__(*bitmaps)
```
```python
return bitmaps[0].copy()
```

Ezibenroc commented on pyroaring/abstract_bitmap.pxi:
```diff
@@ -222,16 +309,17 @@ cdef class AbstractBitMap:
-            return cls(*bitmaps)
+            return bitmaps[0].__class__(*bitmaps)
+        elif size == 2:
+            return (<AbstractBitMap>bitmaps[0]).binary_op(<AbstractBitMap?>bitmaps[1], croaring.roaring_bitmap_or)
```
```python
return bitmaps[0] | bitmaps[1]
```

Ezibenroc commented on pyroaring/abstract_bitmap.pxi:
```diff
@@ -241,12 +329,14 @@ cdef class AbstractBitMap:
         cdef AbstractBitMap result, bm
         if size <= 1:
-            return cls(*bitmaps)
+            return bitmaps[0].__class__(*bitmaps)
```
```python
return bitmaps[0].copy()
```

Ezibenroc commented on pyroaring/abstract_bitmap.pxi:
```diff
@@ -241,12 +329,14 @@ cdef class AbstractBitMap:
-            return cls(*bitmaps)
+            return bitmaps[0].__class__(*bitmaps)
+        elif size == 2:
+            return (<AbstractBitMap>bitmaps[0]).binary_op(<AbstractBitMap?>bitmaps[1], croaring.roaring_bitmap_and)
```
```python
return bitmaps[0] & bitmaps[1]
```

Ezibenroc commented on pyroaring/bitmap.pxi:
```diff
@@ -116,6 +116,64 @@ cdef class BitMap(AbstractBitMap):
+        BitMap([3, 4, 10])
+
+        """
+        union = self.__ixor__(other)
```
You don't need to store the result (which is `None`), just use this:
```python
self.__ixor__(other)
```

Ezibenroc commented:
Thanks a lot for your work!
